### ISSUE TITLE
Rule idea: no imprecise number literals

### ISSUE THREAD

**Issue by michaelficarra** [eslint/eslint#8447]
**Please describe what the rule should do:**

Flag number literals which do not precisely represent an IEEE754 double. Have an option to allow number literals that round to `Infinity`, such as `2e308`, a somewhat common pattern to avoid referring to the `Infinity` global.

**What category of rule is this? (place an "X" next to just one item)**

- [ ] Enforces code style
- [X] Warns about a potential error
- [ ] Suggests an alternate way of doing something
- [ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
0.1234567890123456789
```

```js
12345678901234567890
```

```js
12345678.901234567890
```

```js
0xFFFFFFFFFFFFFFF
```

**Why should this rule be included in ESLint (instead of a plugin)?**

It's a possible error that can affect all JavaScript users and can lead to logic bugs that are very hard to notice.

**Comment by rightfold** [comment]
What about `0.1`? It is not exactly one tenth:
```
> 0.1.toFixed(20)
"0.10000000000000000555"
```

**Comment by michaelficarra** [comment]
Yep, `0.1`, `0.2`, `0.3`, all imprecise.

**Comment by rightfold** [comment]
The warning should probably show the exact value of the literal, so that it can be substituted for by the programmer.

**Comment by not-an-aardvark** [comment]
The original issue makes sense to me (it doesn't seem like a good idea to have literals with extra precision that doesn't get used), but I'm confused by the following discussion about `0.1`. Wouldn't it be impossible to represent `0.1` precisely with JS numbers?

It seems like the error message for something like `0.1234567890123456789` would be "Use 0.12345678901234568 instead", whereas the error message for `0.1` would have to be "don't use this value at all". Is the goal of the rule to disallow unused precision info in number literals in favor of stating the actual value explicitly, or is it to disallow the use of imprecise values at all?

**Comment by michaelficarra** [comment]
@not-an-aardvark The latter. Over-precise literals are a subset of literals which do not describe a double. For example, `9007199254740992` and `9007199254740994` are both precisely representable, but the integer between them (`9007199254740993`) is not, and gets rounded to `9007199254740992`. This rule should flag `9007199254740993`.

**Comment by michaelficarra** [comment]
@not-an-aardvark And in the `0.1` case, you should instead write `0.10000000000000000555`, a representation of the real value you're working with if you type `0.1`.

**Comment by not-an-aardvark** [comment]
In practice, what's the advantage of referring to `0.10000000000000000555` rather than `0.1`? Does the system become more mathematically sound (e.g. by avoiding unexpected rounding errors)?

To me, `0.1` seems more readable than `0.10000000000000000555` because the former makes it clear that the literal is supposed to represent the mathematical value of `1/10`, to the extent that floating-point precision can allow.

**Comment by ilyavolodin** [comment]
@michaelficarra I assume this is done for internal optimization of the code and to avoid rounding errors. Could you provide a real-world example where this rule would be useful? I'm just having a very hard time coming up with anything specific where I would need this kind of precision.

**Comment by michaelficarra** [comment]
@not-an-aardvark It's about setting expectations. If a program contains a numeric literal that looks like one mathematical number but actually represents a different mathematical number, that program is deceptive. A reader may not have memorised every single IEEE754 double, so may not know what number is being represented when the literal `9007199254740993` is used. The program would be more clearly written using `9007199254740992`.

@ilyavolodin It's not about performance or optimisation at all. How many times does the following loop body get evaluated?

```js
for (let i = 0; i <= 0.3; i += 0.1)
  console.log(i);
```

4, right? Nope. Now if it had been written like this, you would be able to better predict that it is only evaluated 3 times.

```js
for (let i = 0; i <= 0.2999999999999999889; i += 0.10000000000000000555)
  console.log(i);
```

Of course, this is a very contrived example, but I would argue that it is never better to write out a mathematical value that will be immediately rounded to another mathematical value when the program is loaded into memory.

**Comment by ilyavolodin** [comment]
Wow, you learn something new every day. I didn't realize that your first example would only loop 3 times. Interesting. I don't mind adding a rule like that. I do think that there are very few people who will find it useful, though, but I think that's fine.

**Comment by not-an-aardvark** [comment]
This seems extremely situational and I think it would be inconvenient in the common case, so I'm :-1: for it.

**Comment by platinumazure** [comment]
I like the idea in general, but I think this fits better in a plugin.

**Comment by michaelficarra** [comment]
@not-an-aardvark Most of the numbers you use in your programs (small integers) are precisely representable. The common case is no warnings.

**Comment by not-an-aardvark** [comment]
To clarify, I was referring to the "common case" for programs that use non-integer values. (This might be a small subset of programs in general, but I think programs where floating-point precision actually matters is a small subset of programs that use non-integer values.)

Also, I assume the rule wouldn't be able to catch expressions like `1/10`, which could be misleading in the same way because they don't precisely represent the mathematical value `0.1`. So I think this probably isn't a good fit for core because:

* The problem that the rule tries to address is relatively rare
* The rule would only able to detect some cases where the problem is present
* The rule would have a lot of false positives (it would warn in cases where there isn't actually a precision bug), and would suggest a fix which is less readable than the original code.

**Comment by platinumazure** [comment]
I don't think this is a good fit for core, either. I would love to see a plugin for this which does all of the precision/rounding calculations, but I wouldn't want to take that on in core.

In addition, I agree with most of the reasoning in [this comment](https://github.com/eslint/eslint/issues/8447#issuecomment-293765941). I think if we implement this in core, we would get a lot of complaints about usability. (Compare `no-magic-numbers`.)

So I'm going to :-1: this as well.

**Comment by ilyavolodin** [comment]
Considering this issue has been opened for over a year, and it does have 2 downvotes from the team, I think it's time to close it. While I think everyone agrees that the proposal covers correct behavior, it's extremely situational and would only apply to a few people (I'm not sure why would anyone write loops with fractions, and the suggested solution for fixing imprecision in the loop declaration is unreadable to most people who don't do fractional math on daily bases). So I'm going to close this issue. If you disagree with this, please re-open it.
